### PR TITLE
Fixed cropping polygon in some corner cases

### LINKED ISSUE #3178: Polygon cropping bug
### My actions before raising this issue
- [x] Read/searched [the docs](https://github.com/opencv/cvat/tree/master#documentation)
- [x] Searched [past issues](/issues)

There is the bug while cropping polygon outside the border of images.

### Expected Behaviour
Polygons should be cropped properly.

### Current Behaviour
Polygons are cropped in unexpected ways.

### Steps to Reproduce (for bugs)
Seems that this issue depends on direction of drawing. On the right part of the image the issues occurs while you draw it in clockwise direction and on the left part of the image with counter-clockwise direction respectively.

Firstly I draw the polygon...
cvat_crop_bug_1

... and than I press "N" hotkey for finish this polygon.

cvat_crop_bug_2

### Your Environment

- Git hash commit (`git log -1`):
```
commit 6a0379c3b9163f2d5c66b787f67015648fca88e9 (HEAD -> develop, origin/develop, origin/HEAD)
Author: Boris Sekachev <boris.sekachev@intel.com>
Date:   Fri May 7 16:30:16 2021 +0300

    Fixed hundreds of vulnerabilities in npm audit (#3169)
    
    * Fixed hundreds of vulnerabilities in npm audit
    
    * Updated package-lock for canvases
    
    * Fixed package-lock
    
    * Fixed vulnerabilities in cvat-canvas3d and cvat-data

```
- Docker version `docker version` (e.g. Docker 17.0.05): 20.10.05
- Are you using Docker Swarm or Kubernetes? `No`
- Operating System and version (e.g. Linux, Windows, MacOS): GNU/Linux Ubuntu 4.15.0-140-generic

### Update
Here are my settings (but this issue does not depend on them).
cvat_crop_bug_settings

### PR DESCRIPTION
**Pull request by bsekachev**
### Motivation and context
Resolved #3178

Hi @dvkruchinin 
A test for this case would be good. 

### How has this been tested?
Manual testing

### Checklist

- [x] I submit my changes into the `develop` branch
- [x] I have added description of my changes into [CHANGELOG](https://github.com/opencv/cvat/blob/develop/CHANGELOG.md) file
- [ ] I have updated the [documentation](
  https://github.com/opencv/cvat/blob/develop/README.md#documentation) accordingly
- [ ] I have added tests to cover my changes
- [x] I have linked related issues ([read github docs](
  https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword))
- [x] I have increased versions of npm packages if it is necessary ([cvat-canvas](https://github.com/opencv/cvat/tree/develop/cvat-canvas#versioning),
[cvat-core](https://github.com/opencv/cvat/tree/develop/cvat-core#versioning), [cvat-data](https://github.com/opencv/cvat/tree/develop/cvat-data#versioning) and [cvat-ui](https://github.com/opencv/cvat/tree/develop/cvat-ui#versioning))

### License

- [x] I submit _my code changes_ under the same [MIT License](
  https://github.com/opencv/cvat/blob/develop/LICENSE) that covers the project.
  Feel free to contact the maintainers if that's a concern.
- [x] I have updated the license header for each file (see an example below)

```python
# Copyright (C) 2021 Intel Corporation
#
# SPDX-License-Identifier: MIT
```

### CODE REVIEW

File: cvat-canvas/src/typescript/drawHandler.ts [hunk 1]
```diff
@@ -121,22 +123,35 @@ export class DrawHandlerImpl implements DrawHandler {
...
+                resultPoints.push(...findInersection(p1, p2, bottomLine[0], bottomLine[1]));
+                resultPoints.push(...findInersection(p1, p2, topLine[0], topLine[1]));
             }
 
             if (resultPoints.length === 4) {
                 if (
-                    Math.sign(resultPoints[0] - resultPoints[2]) !== Math.sign(p1.x - p2.x)
-                    && Math.sign(resultPoints[1] - resultPoints[3]) !== Math.sign(p1.y - p2.y)
+                    (p1.x === p2.x || Math.sign(resultPoints[0] - resultPoints[2]) !== Math.sign(p1.x - p2.x))
+                    && (p1.y === p2.y || Math.sign(resultPoints[1] - resultPoints[3]) !== Math.sign(p1.y - p2.y))
```
Comment (bsekachev) [inline cvat-canvas/src/typescript/drawHandler.ts:154]: @azhavoro  @ActiveChooN 

The core change to fix (others are to satisfy eslint, minor refactoring, and some minor bug fixes). 

As I understand the issue was following:
After the first (Horizontal) crop p1.x in a corner case could be equal to p2.x).
The same if we reverse crop order (first is Vertical), then p1.y in a corner case could be equal to p2.y.

In these corner cases the condition can be false and we do not reverse points when it is actually necessary.

**Comment by dvkruchinin** [comment]
> A test for this case would be good.

Sure. A test will be prepared.

File: cvat-canvas/src/typescript/drawHandler.ts [hunk 2]
```diff
@@ -238,27 +245,34 @@ export class DrawHandlerImpl implements DrawHandler {
 
         if (cuboidOffsets.length === points.length / 2) {
             cuboidOffsets.forEach((offsetCoords: number[]): void => {
-                if (Math.sqrt((offsetCoords[0] ** 2) + (offsetCoords[1] ** 2)) < minCuboidOffset.d) {
-                    minCuboidOffset.d = Math.sqrt((offsetCoords[0] ** 2) + (offsetCoords[1] ** 2));
+                const dx = offsetCoords[0] ** 2;
+                const dy = offsetCoords[1] ** 2;
```
Comment (bsekachev) [inline cvat-canvas/src/typescript/drawHandler.ts:249]: Conflict between prettier and eslint. 
There were suggestion to split to several lines: 
https://github.com/prettier/eslint-config-prettier#no-mixed-operators

File: cvat-canvas/src/typescript/drawHandler.ts [hunk 3]
```diff
@@ -96,9 +96,9 @@ export class DrawHandlerImpl implements DrawHandler {
             Vertical,
         }
 
-        const isBetween = (x1: number, x2: number, c: number): boolean => (
-            c >= Math.min(x1, x2) && c <= Math.max(x1, x2)
-        );
+        function isBetween(x1: number, x2: number, c: number): boolean {
+            return c >= Math.min(x1, x2) && c <= Math.max(x1, x2);
+        }
```
Comment (bsekachev) [inline cvat-canvas/src/typescript/drawHandler.ts:101]: Conflict between prettier and eslint. The first removes parentness, the second removes extra line and gets too many characters in the line (> 120) => error. No ways to configure prettier ((

Comment (bsekachev) [reply cvat-canvas/src/typescript/drawHandler.ts:101]: One way to solve is to transform arrow function to general function

**Review by ActiveChooN — APPROVED** [review]
LGTM

**Review by azhavoro — APPROVED** [review]
